### DISCUSSION TITLE
Usefull gcode macroses/editions

### DISCUSSION THREAD

**Discussion by g992** [xblax/flashforge_ad5m_klipper_mod#7] [General]
I tried to make it more convenient and accessible, like stock software. Therefore, I suggest sharing macros/settings here that could make your experience easier
1. Changing the parameters of the roaming, it is very important for us to lower the table first, then park X and then Y, since the magnets are located in one place
```
[homing_override]
gcode:
    G28 Z0
    G28 X0
    G28 Y0
```
2. Loading the table level profile and setting the offset after launch
```
[delayed_g code bed_mesh_init]
initial_duration: .01
gcode:
  BED_MESH_PROFILE LOAD=default
  SET_GCODE_OFFSET Z=0.325
```
3. I tried to edit the pause script, but I couldn't, it is necessary to set the sequence Z-X-Y

**Comment by xblax** [comment]
Homing:
Why does the order of homing matters. Can you elaborate more? I also don't quite understand how the homing works, I don't see switches at least. Did you see the changes I commited to printer.base.cfg yesterday? That seems to make Homing much more accurate.
Edit: if not configured otherwise, the order of Axis seems to be X, Y, then Z.

Leveling:
I would prefer to not load a default profile automatically. My idea is to do automatic leveling on print start, but only of no mesh is loaded. So if you do multiple prints in a row it levels only once. We also need a nozzle wipe on the bed before leveling like the stock software does, to make sure that it is clean from Filament. Where does the Gcode offset come from? That shouldn't be needed.

I'm currently experimenting with the PROBE_ACCURACY command in Klipper. Stock settings give bad accuracy, but it seems it can be improved.

**Reply by jlodew0** [reply]
I do like the implementation of the default wipe as well. First heat up build plate and nozzle to temp. extrude 3 small lines. Keep the nozzle on the build Plate and let it cool down to 120 degrees so no oozing happens and then probe the build Plate. I came from a modified ender 3 pro. So I’m still easy to impress I guess :D

**Comment by xblax** [comment]
I did a trace on Kipper unix socket to capture Gcodes sent by Flashforge software print start process including bed leveling. I think that is a good starting point to optimize the Macros.

[flashforge_gcode_print_start.log](https://github.com/xblax/flashforge_adm5_klipper_mod/files/14352089/flashforge_gcode_print_start.log)

**Comment by g992** [comment]
> Homing:
> Why does the order of homing matters. Can you elaborate more?
The sequence Z-X-Y is important to us because:
1. The printer has corners on the table, if we try to go to 0-0-0 and the nozzle is at the table level, and the first axis is NOT Z, then we will rest the nozzle against the corner for fixing the pei plate.
2. The printer has built-in magnets in the right rear corner to activate the hall sensors installed in the print head. if the printer suddenly parks from the center of the table, the Y terminal will not work, but it does not apply to the x-y-z sequence.

**Reply by xblax** [reply]
Thanks, will add a homing override as suggested. Tried homing Y before X yesterday and found out the hard way. But how does X-homing work? I couldn't find any magnet on the gantry.

**Reply by g992** [reply]
idk, magnet inside carriage Y? i check this theory at home
![telegram-cloud-photo-size-2-5382024609995544596-x](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/48438685/3c6e67f5-daba-4210-86c2-4b39dc6f00d9)

**Comment by KaruroChori** [comment]
> The printer has built-in magnets in the right rear corner to activate the hall sensors installed in the print head. if the printer suddenly parks from the center of the table, the Y terminal will not work, but it does not apply to the x-y-z sequence.

I am extremely confused, what is this feature about? Why does the print head have a hall sensor inside?

**Reply by jlodew0** [reply]
I would guess as endstop detection. Correct me if I’m wrong @g992 .

I also noticed that the warming up starting from the far right corner, somewhat melts the plastic  tab that holds the pei sheet. So I would (maybe change my start g-code) to move a bit away of that corner or drop the z a bit before warming up. However, @xblax was looking into what ff does default for wiping the nozzle before probing. Maybe we can add that to the section of specific gcodes and document why we do certain things? (In between gcodes)

**Reply by g992** [reply]
printhead have 2 hall sensors for x and y
<img src="https://github.com/xblax/flashforge_adm5_klipper_mod/assets/48438685/e2525854-cf54-4036-817a-45a805ff6a84" width="200" />

**Reply by xblax** [reply]
I think homing override also should also move the head and bed away from homing position by 5mm on each axis. Noticed the same issue with the tab that holds the sheet.

I did came up with a nozzle wipe macro for automatic bed leveling routine yesterday, but will follow-up with this later today. 

I think the whole auto-leveling will work quite nice. It's should not be possible to start a print without having a mesh loaded, because it will never work. For me, the distance to the bed before leveling is around 2.5mm, I assume it's similar for others.

**Comment by g992** [comment]
Important: edit the pause script so that the printer first raises the z, as if the pause occurs on the first layers. the corner of the bed is damaged

**Comment by AJolly** [comment]
Probably want to create a [safe_z_home] /  [homing_override] script .  Also since we have spi access to the stepper drivers we can enable [endstop_phase] calibration.  I get my 5m tonight :)

**Comment by xblax** [comment]
I pushed updated macros / printer configuration yesterday: 

- I hope the changes now cover all the all the safety issues to stay away far enough away from the plastic corner.
- Main improvements are for bed leveling and automatic leveling on print start:
  - `LOAD_CELL_TARE`
This command resets the sensor readout of the load cell to 0, automatically done before leveling - usually not needed manually unless doing manual `PROBE` or `PROBE_ACCURACY`. I implemented this similar to Flashforge stock software. Flashforge does automatic retry on failure, up to 10 times, but I don't now how to do that with Klipper Macros.
  - `AUTO_BED_LEVEL` 
This does an automatic bed leveling routine with nozzle wipe and cooldown to 120° before doing the mesh bed calibration. 
Parameters are `BED_TEMP=[60] EXTRUDER_TEMP=[200] PROFILE=[auto]`
  - `START_PRINT` does now automatically do `AUTO_BED_LEVEL` before print if no mesh is loaded. 
Also added parameters `FORCE_LEVELING` and `DISABLE_PRIMING`

Only issue I have is that the Nozzle wipe moves don't really remove filament sticking to the nozzle. Maybe we should do an extrusion similar to nozzle priming instead. But anyway, it makes sure that the nozzle presses flat again the bed. In first tests I seem to get really good first layers, no manual offset or flow-rate adjustment needed.

**Comment by g992** [comment]
Some words before release:
1. on all versions, after start printing we need to enable pcb fan  that on PB7, because tmc2209 can die
2. on PRO versions, need enable led light after start, just  initial_WHITE: 0.8
3. on PRO, need to create macross that enable or switch + enable filter fan

**Comment by xblax** [comment]
@g992:
PCB fan is already enabled automatically in the latest Macros.

Will add enabling the Led in general, as I'm also planning to add LEDs to my 5M - I think that will be a quite common mod. 
https://www.thingiverse.com/thing:6358773

For PRO specific Macros. I don't have the 5M Pro, so I hope that's what someone can contribute in a pull request later on. I don't think that's critical for a public beta.
